src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java

```diff
@@ -604,7 +606,56 @@ private int implUpdate(byte[] in, int inOfs, int inLen,
             ensureInitialized();
             int k = 0;
             int newPadBufferLen = 0;
-            if (paddingObj != null && (!encrypt || reqBlockUpdates)) {
+            if (blockMode == Mode.CTS) {
+                // decide how to split the total data (totalInLen) between
+                // the token (dataForP11Update) and padBuffer
+                // (newPadBufferLen)
+                int totalInLen = padBufferLen + inLen;
+                newPadBufferLen = getCTSMustBeBuffered(totalInLen);
+                int dataForP11Update = totalInLen - newPadBufferLen;
+                if (dataForP11Update > 0 && padBufferLen > 0) {
+                    // there is data for the token and part of it is in
+                    // padBuffer
+                    int flushFromPadBuffer;
+                    int fillLen = getBytesToCompleteBlock(padBufferLen);
+                    if (dataForP11Update >= padBufferLen + fillLen) {
```

Comment:
Maybe use `if (inLen >= fillLen)` ?

Reply:
`dataForP11Update >= padBufferLen + fillLen` is not the same as `inLen >= fillLen` (the equivalent would be `inLen - newPadBufferLen >= fillLen`, but I personally find the proposed condition more clear). We will flush the entire `padBuffer` only if there are enough bytes in `inLen` to fill `padBuffer` with whatever we need (0 or more bytes) and fulfill the new buffering requirement. Regarding `fillLen > 0`, that is not strictly needed to flush the entire `padBuffer`. If we are buffering 3 blocks (e.g. for NSS), we may have 1 block buffered in `padBuffer` and `fillLen == 0` (no need to borrow to complete `padBuffer` to a multiple of a block size).

Reply:
I see, thanks for the explanation.